File: Backend/src/controller/usercontroller/loginController.ts
import { NextFunction, Request, Response } from 'express';
import jwt from 'jsonwebtoken';
import User, { IUser } from '../../Database/models/userModel';
import bcryptjs from 'bcryptjs';
import { userSchemaValidationLogin } from '../../utils/JoiUtils/joiLoginValidation';
import createHttpError from 'http-errors';
import { config } from '../../config/config';

const loginuser = async (
    req: Request,
    res: Response,
    next: NextFunction
): Promise<Response | void> => {
    try {
        // Fetch all user data from req.body
        const { username, email, password } = req.body;

        // Check if all required fields are provided in the request body
        if (!username || !email || !password) {
            return next(createHttpError(400, 'All fields are required'));
        }

        // Validate user data using JOI
        const { error } = userSchemaValidationLogin.validate(req.body);
        if (error) {
            return next(createHttpError(400, error.details[0].message));
        }

        // Find the user in the database
        const checkUser: IUser | null = await User.findOne({
            username,
            email,
        });
        if (checkUser) {
            // Match password
            const matchPassword = await bcryptjs.compare(
                password,
                checkUser.password
            );
            if (matchPassword) {
                // Create JWT Token
                const userPayload = {
                    fullname: checkUser.fullname,
                    email: checkUser.email,
                    username: checkUser.username,
                    gender: checkUser.gender,
                    country: checkUser.country,
                    currency_code: checkUser.currency_code,
                    _id: checkUser._id,
                };

                // Access token
                const accessToken = jwt.sign(
                    userPayload,
                    process.env.JWT_ACCESS_SECRET as string,
                    { expiresIn: '1h' }
                );
                // Refresh token
                const refreshToken = jwt.sign(
                    userPayload,
                    process.env.JWT_REFRESH_SECRET as string,
                    { expiresIn: '7d' }
                );

                // Save the refresh token into the database for future use
                checkUser.refreshtoken = refreshToken;
                await checkUser.save();

                // Set cookies
                res.cookie('accessToken', accessToken, {
                    httpOnly: true,
                    secure: process.env.NODE_ENV === 'production',
                    sameSite: 'strict',
                });

                return res.status(200).send({
                    status: 'Success',
                    accessToken: accessToken,
                    refreshToken: refreshToken,
                });
            } else {
                return next(createHttpError(401, 'Incorrect Password'));
            }
        } else {
            return next(createHttpError(404, 'User not found!'));
        }
    } catch(error) {
        if (config.env === 'development') {
            console.error('Error during login:', error); // Log for debugging
        }
        return next(createHttpError(500, 'Internal Server Error!'));
    }
};

export default loginuser;

File: APIs/app/modules/users/controllers/login_controller.py
from fastapi import Response, HTTPException, status
from app.core.database.models.user_model import User
from app.core.utils.validation.login_validation import LoginRequest
from app.core.config.config import config
import jwt
from passlib.context import CryptContext
from datetime import datetime, timedelta

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

async def login_user(
    request: LoginRequest,
    response: Response
):
    # Find user
    check_user = await User.find_one({
        "username": request.username,
        "email": request.email
    })
    
    if check_user:
        # Match Password
        if check_user.password and pwd_context.verify(request.password, check_user.password):
            # Create Token Payload
            user_payload = {
                "fullname": check_user.fullname,
                "email": check_user.email,
                "username": check_user.username,
                "gender": None, # Not in schema but in TS payload, default None
                "country": check_user.country,
                "currency_code": check_user.currency_code,
                "_id": str(check_user.id)
            }

            # Generate tokens
            access_token = jwt.encode(
                {**user_payload, "exp": datetime.utcnow() + timedelta(hours=1)},
                config.JWT_ACCESS_SECRET,
                algorithm="HS256"
            )
            
            refresh_token = jwt.encode(
                {**user_payload, "exp": datetime.utcnow() + timedelta(days=7)},
                config.JWT_REFRESH_SECRET,
                algorithm="HS256"
            )

            # Save refresh token (Beanie save)
            check_user.refreshtoken = refresh_token
            await check_user.save()

            # Set Cookies
            response.set_cookie(
                key="accessToken",
                value=access_token,
                httponly=True,
                secure=config.ENV == "production",
                samesite="strict"
            )

            return {
                "status": "Success",
                "accessToken": access_token,
                "refreshToken": refresh_token
            }
        else:
            raise HTTPException(status_code=401, detail="Incorrect Password")
    else:
        raise HTTPException(status_code=404, detail="User not found!")
